rust/src/derivation.rs

```diff
@@ -374,6 +376,130 @@ impl zcash_address::TryFromRawAddress for UnifiedAddressParser {
     }
 }
 
+/// A struct that contains a Zcash unified address, along with the diversifier index used to
+/// generate that address.
+#[repr(C)]
+pub struct FfiAddress {
+    address: *mut c_char,
+    diversifier_index_bytes: [u8; 11],
+}
+
+impl FfiAddress {
+    fn new(
+        network: &Network,
+        address: UnifiedAddress,
+        diversifier_index: DiversifierIndex,
+    ) -> Self {
+        let address_str = address.encode(network);
+        Self {
+            address: CString::new(address_str).unwrap().into_raw(),
+            diversifier_index_bytes: *diversifier_index.as_bytes(),
+        }
+    }
+}
+
+/// Frees a FfiAddress value
+///
+/// # Safety
+///
+/// - `ptr` must be non-null and must point to a struct having the layout of [`FfiAddress`].
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_free_ffi_address(ptr: *mut FfiAddress) {
+    if !ptr.is_null() {
+        let ffi_address: Box<FfiAddress> = unsafe { Box::from_raw(ptr) };
+        if !(ffi_address.address.is_null()) {
+            unsafe { zcashlc_string_free(ffi_address.address) }
+        }
+        drop(ffi_address);
+    }
+}
+
+/// Derives a unified address address for the provided UFVK, along with the diversifier at which it
```

Comment:
Would we ever need a `zcashlc_derive_uivk_from_ufvk`? I ask because that's not expressible using these two functions.

Reply:
We should likely expose it, yes; the derivation tool should provide this sort of functionality.

Reply:
Opened https://github.com/Electric-Coin-Company/zcash-light-client-ffi/issues/175